src/components/common/dialog.tsx

```diff
@@ -0,0 +1,1550 @@
+    }
+
+    /* The first thing should be of course to probe that cache.  If we
+       get a hit, it is used immediately to call #set_validation.
+
+       In that case, the DialogValidationState is also made part of
+       the current round since any asynchronous validation that is
+       currently running is still relevant. See below for more about
+       that.
+     */
+
+    #probe_validation_state_cache(state: DialogValidationState, val: unknown): boolean {
+        if (Object.is(state.cached_value, val)) {
+            state.round_id = this.#get_current_validation_round_id();
+            debug("cache hit", state.path, state.cached_result);
+            this.#set_validation(state.path, state.cached_result);
+            return true;
+        } else
+            return false;
+    }
+
+    /* And in fact, _validate_value does exactly those two things.
+     */
+
+    _validate_value(path: string, val: unknown, func: () => string | undefined): void {
+        const state = this.#get_validation_state(path);
+        if (!this.#probe_validation_state_cache(state, val)) {
+            const result = func();
+            debug("sync validate", state.path, result);
+            this.#set_validation_state_result(state, val, result);
+        }
+    }
+
+    /* Now asynchronous validation.
+
+       Each call to #trigger_validation starts a new "validation
+       round" and a DialogValidationState keeps track to which round
+       it applies to.  This matters of course for asynchronous
+       validation: If async validation for a given path was started in
+       one round, and then the next round happens but the path is no
+       longer enumerated by the validation callback (i.e., its value
+       is no longer relevant for the dialog), then this asynchronous
+       validation should have no effect.
+
+       We use the #validation object as the round identifier, since it
+       is created fresh by each call to #trigger_validation.
+     */
+
+    #get_current_validation_round_id(): unknown {
+        return this.#validation;
+    }
+
+    #is_current_validation_round_id(id: unknown): boolean {
+        return Object.is(id, this.#validation);
+    }
+
+    /* If there was no cache hit, asynchronous validation starts with
+       a timeout, followed by letting a asynchronous function run to
+       resolution.
+
+       Setting a new timeout of course cancels any previously set
+       one. It also installs the current value in the cache, so that
+       subsequent validation rounds do nothing until the value
+       actually changes.
+
+       One interesting thing to note is that when doing the final
+       validation before running an action function, no debouncing
+       delay should be applied of course. We want to get on with
+       validation immediately.
+     */
+
+    #set_validation_state_timeout(
+        state: DialogValidationState,
+        val: unknown,
+        delay: number,
+        func: () => void,
```

Comment:
These 5 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2425-09c34802-20251217-183240-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)